web/src/typeahead_helper.ts

```diff
@@ -557,6 +558,151 @@ export function sort_recipients<UserType extends UserOrMentionPillData | UserPil
+
+        if (user_group_a.is_system_group && !user_group_b.is_system_group) {
+            return -1;
+        }
+
+        if (user_group_b.is_system_group && !user_group_a.is_system_group) {
+            return 1;
+        }
+
+        if (user_group_a.name < user_group_b.name) {
+            return -1;
+        }
+
+        return 1;
+    }
+
+    assert(option_a.type === "user");
+    assert(option_b.type === "user");
+
+    if (target_group !== undefined) {
+        if (
+            !target_group.members.has(option_a.user.user_id) &&
+            target_group.members.has(option_b.user.user_id)
+        ) {
+            return 1;
+        }
+
+        if (
+            target_group.members.has(option_a.user.user_id) &&
+            !target_group.members.has(option_b.user.user_id)
+        ) {
+            return -1;
+        }
+    }
+
+    if (option_a.user.full_name < option_b.user.full_name) {
+        return -1;
+    } else if (option_a.user.full_name === option_b.user.full_name) {
+        return 0;
+    }
+
+    return 1;
+}
+
+export function sort_group_setting_options({
+    users,
+    query,
+    groups,
+    target_group,
+}: {
+    users: UserPillData[];
+    query: string;
+    groups: UserGroupPillData[];
+    target_group: UserGroup | undefined;
+}): (UserPillData | UserGroupPillData)[] {
+    function sort_group_setting_items(
+        objs: (UserPillData | UserGroupPillData)[],
+    ): (UserPillData | UserGroupPillData)[] {
+        objs.sort((option_a, option_b) =>
+            compare_setting_options(option_a, option_b, target_group),
+        );
+        return objs;
+    }
+
+    const users_name_results = typeahead.triage_raw(query, users, (p) => p.user.full_name);
+    const email_results = typeahead.triage_raw(
+        query,
+        users_name_results.no_matches,
+        (p) => p.user.email,
+    );
+    const groups_results = typeahead.triage_raw(query, groups, (g) => g.name);
+
+    const exact_matches = sort_group_setting_items([
+        ...groups_results.exact_matches,
+        ...users_name_results.exact_matches,
+        ...email_results.exact_matches,
+    ]);
+
+    const prefix_matches = sort_group_setting_items([
+        ...groups_results.begins_with_case_sensitive_matches,
+        ...groups_results.begins_with_case_insensitive_matches,
+        ...users_name_results.begins_with_case_sensitive_matches,
+        ...users_name_results.begins_with_case_insensitive_matches,
+        ...email_results.begins_with_case_sensitive_matches,
+        ...email_results.begins_with_case_insensitive_matches,
+    ]);
+
+    const word_boundary_matches = sort_group_setting_items([
+        ...groups_results.word_boundary_matches,
+        ...users_name_results.word_boundary_matches,
+        ...email_results.word_boundary_matches,
+    ]);
+
+    const no_matches = sort_group_setting_items([
+        ...groups_results.no_matches,
+        ...email_results.no_matches,
```

Comment:
The typeahead is sorted in following order -
- exact match
- prefix match without any preference given whether the case matches or not.
- word boundary match
- no matches

And the each list above is sorted in order system groups > groups > users who are members of group (for group setting) > other users. We do not have group creator data now (the PR for that is close to being completed), so can add that later.

Reply:
Playing with the PR, let's drop the exact mach distinction -- I think we should treat matches the same regardless of case.

I'm not sure if we'll end up wanting to prioritize groups more strongly, but it'll be easier to play with that after we make this change.

Reply:
Showing bots mixed in with users feels off, so let's also put all humans above all bots. We can do:

users who are members of group > other users > bots who are members of the group > other bots

Reply:
The exact match distinction is not for case, it matches only when entire name of user/group is present in the input like below irrespective whether case matches or not.

![Screenshot from 2024-09-07 12-09-14](https://github.com/user-attachments/assets/9c36c539-912e-4eba-b3f2-ab55defd6cd4)

Reply:
And we do not show bots at all in the typeahead.

Reply:
Oh, I got confused somehow. It was seeing the bots mixed up with people in group membership that felt off. We can tackle that separately.

Reply:
But why do I see this? I would expect Moderators to be ranked higher.

![Screenshot 2024-09-09 at 15 40 45](https://github.com/user-attachments/assets/18f956f4-e8b3-4ee1-8ae9-63ed38e4829e)

Reply:
Fixed this -
![Screenshot from 2024-09-10 21-39-35](https://github.com/user-attachments/assets/c32bd9ab-4153-4ce5-89df-fe3f25441015)